associated-token-account/program/src/processor.rs

```diff
@@ -56,6 +72,26 @@ pub fn process_create_associated_token_account(
     let spl_token_program_info = next_account_info(account_info_iter)?;
     let spl_token_program_id = spl_token_program_info.key;
 
+    if create_mode == CreateMode::Idempotent {
+        if associated_token_account_info.owner == spl_token_program_id {
+            let ata_data = associated_token_account_info.data.borrow();
+            if let Ok(associated_token_account) = StateWithExtensions::<Account>::unpack(&ata_data)
+            {
+                if associated_token_account.base.owner != *wallet_account_info.key {
+                    let error = AssociatedTokenAccountError::InvalidOwner;
+                    msg!("{}", error);
+                    return Err(error.into());
+                }
+                if associated_token_account.base.mint != *spl_token_mint_info.key {
+                    return Err(ProgramError::InvalidAccountData);
+                }
+                return Ok(());
+            }
+        } else if *associated_token_account_info.owner != system_program::id() {
+            return Err(ProgramError::IllegalOwner);
+        }
```

Comment:
I dunno, do we really need this? And if so, why is it specific to CreateMode::Idempotent?
Aren't `associated_token_account_info.owner != system_program::id()` cases just going to error out in `create_pda_account()` when the invokes are attempted?

Reply:
It's more about making the check clear in the code so future modifications don't introduce errors.  Relying on another method to do the right thing is how wormhole was exploited.

I'd be fine with moving it above the instruction variant check though

Reply:
Yes, well, if the System Program breaks that badly, I'm not sure this will be the biggest of our problems.

Reply:
Your call, @joncinque . But I think it needs to be outside the instruction variant case, though (and after seems less of a pita than before)

Reply:
It costs very little to add in outside this check, so I'll move it